File: src/main/java/ir/ut/ai/search/SearchBase.java
package ir.ut.ai.search;

import ir.ut.ai.base.Action;
import ir.ut.ai.base.Node;
import ir.ut.ai.base.Problem;

import java.util.ArrayList;
import java.util.List;

public abstract class SearchBase<S, A> {

    protected Problem<S, A> problem;

    public SearchBase(Problem<S, A> problem) {
        this.problem = problem;
    }

    protected Node<S, A> nexNode(Node<S, A> parentNode, Action<A> action) {
        Node<S, A> resultedNode = problem.result(parentNode, action);
        if (resultedNode != null) {
            double pathCost = parentNode.getPathCost() + problem.stepCost(parentNode, action);
            Node<S, A> node = new Node(resultedNode.getState(), parentNode, action, pathCost);
            return node;
        }
        return null;
    }

    protected List<Node<S, A>> solution(Node<S, A> node) {
        List<Node<S, A>> nodes = new ArrayList();
        Node tempNode = node;
        while(tempNode != null) {
            nodes.add(tempNode);
            tempNode = tempNode.getParent();
        }
        return nodes;
    }

    public abstract List<Node<S, A>> run();
}

File: src/main/kotlin/ir/ut/ai/search/SearchBase.kt
package ir.ut.ai.search

import ir.ut.ai.base.Action
import ir.ut.ai.base.Node
import ir.ut.ai.base.Problem

import java.util.ArrayList

abstract class SearchBase<S, A>(protected var problem: Problem<S, A>) {

    protected fun nextNode(parentNode: Node<S, A>, action: Action<A>): Node<S, A>? {
        val resultedNode = problem.result(parentNode, action)
//        val pathCost = parentNode.pathCost + problem.stepCost(parentNode, action)
//        resultedNode?.pathCost = pathCost
        return resultedNode
    }

    protected open fun solution(node: Node<S, A>): List<Node<S, A>> {
        val nodes = ArrayList<Node<S, A>>()
        var tempNode: Node<S, A>? = node
        while (tempNode != null) {
            nodes.add(tempNode)
            tempNode = tempNode.parent
        }
        return nodes
    }

    abstract fun run(): List<Node<S, A>>
}
